### PR TITLE
PLA-406 definition-sync-crash

### PR DESCRIPTION
**Pull request by Allyson-English**
# Proposed Changes
investigating the definition sync crash found that that the `definition.R.DeviceType.Metadatakey` value was returning nil
the proposed solution creates an empty string that is overwritten if the value is not nil, which works locally:
```
		def := *definition.R
		dType := def.DeviceType
		var deviceType string
		if dType != nil {
			deviceType = definition.R.DeviceType.Metadatakey
		}
```

one reason why this solution may not work is that the same value is used as a map key in the section below where the error was initially occurring:
https://github.com/DIMO-Network/device-definitions-api/blob/33dfdad9e00953ac7ff8580c8e5c8a88e3a47c6b/internal/core/commands/sync_dd_search_data.go#L97-L112

other ideas are welcome!

### CODE REVIEW

**Comment by JamesReate** [comment]
Investigating a bit, I think `definition.R.DeviceType.Metadatakey` being nil on your local is more of a config issue - in dev and prod this value was set when the jobs ran. What probably does make sense along your line of thought, is to fail with `.fatal` if the definition.R.DeviceType.Metadatakey is not set. So let's update his PR to just do that.

I just ran the script in prod, and it suceeded fine - which makes me think that at the time it had been failing just due to a timeout b/c elastic was busy or something like that.

**Comment by Allyson-English** [comment]
@JamesReate updated with the changes you proposed above!

**Comment by elffjs** [comment]
I think the crash I saw was this: you can reach [this line](https://github.com/DIMO-Network/device-definitions-api/blob/b22e6a444ad263461d04eeade0783009ecccf546/internal/infrastructure/elastic/elastic_app_search_service.go#L296) in `buildAndExecRequest` with `resp` being nil.

File: internal/infrastructure/elasticsearch/elastic_search_service.go [hunk 1]
```diff
@@ -69,9 +69,10 @@ func (d *ElasticSearch) buildAndExecRequest(method, url string, obj interface{},
 	req.Header.Set("Authorization", "ApiKey "+d.Token)
 	var resp *http.Response
 	var err error
```
Comment (elffjs) [inline internal/infrastructure/elasticsearch/elastic_search_service.go:71]: Does this `err` ever get assigned to before line 106? In particular, it seems like `err == nil` on line 76 will always be true, and `err != nil` will always be false.

File: internal/infrastructure/elasticsearch/elastic_search_service.go [hunk 2]
```diff
@@ -69,9 +69,10 @@ func (d *ElasticSearch) buildAndExecRequest(method, url string, obj interface{},
...
+	var nilRespErr error
 
 	for _, backoff := range backoffSchedule {
-		resp, err = d.httpClient.Do(req) // any error resp should return err per docs
+		resp, nilRespErr = d.httpClient.Do(req) // any error resp should return err per docs
```
Comment (elffjs) [inline internal/infrastructure/elasticsearch/elastic_search_service.go:75]: Could you explain the name a bit?

File: internal/infrastructure/elasticsearch/elastic_search_service.go (hunk 1 above)
Comment (elffjs) [reply internal/infrastructure/elasticsearch/elastic_search_service.go:71]: I think the worst outcome here is that we fail a few times due to, say, timeout; we didn't pass in an `outObj`, and we return with a nil error like everything is time.

File: internal/infrastructure/elasticsearch/elastic_search_service.go [hunk 3]
```diff
@@ -101,9 +102,13 @@ func (d *ElasticSearch) buildAndExecRequest(method, url string, obj interface{},
 		time.Sleep(backoff)
 	}
 	if objOut != nil {
-		err = json.NewDecoder(resp.Body).Decode(&objOut)
-		if err != nil {
-			return nil, errors.Wrap(err, "error decoding response json")
+		if resp != nil {
+			err = json.NewDecoder(resp.Body).Decode(&objOut)
```
Comment (elffjs) [inline internal/infrastructure/elasticsearch/elastic_search_service.go:106]: Hm, is it the caller's responsibility to close the body? It doesn't seem like any of the callers are.

File: internal/core/commands/sync_dd_search_data.go [hunk 4]
```diff
@@ -75,6 +75,13 @@ func (ch SyncSearchDataCommandHandler) Handle(ctx context.Context, query mediato
 			}
 		}
 
+		def := *definition.R
+		dType := def.DeviceType
+		var deviceType string
```
Comment (elffjs) [inline internal/core/commands/sync_dd_search_data.go:80]: Do we want to assign to this? `deviceType := dType.Metadatakey`?

File: internal/core/commands/sync_dd_search_data.go [hunk 5]
```diff
@@ -75,6 +75,13 @@ func (ch SyncSearchDataCommandHandler) Handle(ctx context.Context, query mediato
...
+		if dType == nil {
```
Comment (elffjs) [inline internal/core/commands/sync_dd_search_data.go:81]: It looks like `device_definitions.device_type_id` is still nullable in the database. I wish it weren't, but given this I wouldn't want to tear down the app if it's missing.

Comment (JamesReate) [reply internal/core/commands/sync_dd_search_data.go:81]: ah good point, making a PR to update `device_type_id` to be not null.

**Comment by elffjs** [comment]
@JamesReate We gave rewriting this using `shared` a shot. I think the functionality is the same, would you mind taking a look?

**Review by JamesReate — APPROVED** [review]

**Pull request merged**